### ISSUE TITLE
introduce workspaces

### ISSUE THREAD

**Issue by propensive** [propensive/fury-old#1323]
Every module should have an implicit "workspace", a directory created to act as the module's working directory while an application module is run, and into which the compiler can output classfiles or jarfiles, and which the application module may 

Application modules should be able to freely read and write to their workspace, and downstream modules should be able to depend upon upstream workspaces.

Modules should be able to depend on binaries produced by other modules, however some design work is required to find out a good way to specify how those binaries should be packaged: possible options include,
- the raw classfiles
- classfiles packaged as a JAR
- classfiles packaged as a fat JAR
- classfiles organized into artifacts, which may contain subtrees of modules packaged as JARs, potentially referring to surrogate binaries. (Requires some work in #925)

**Comment by propensive** [comment]
My thinking on this has changed a little.

Workspaces are relevant mainly when running an application module. They should provide read/write space with all the necessary permissions granted for the application to change the contents.

It should be possible to create a workspace associated with a `Layer` in the same namespace as repos, and to declare its contents in terms of artifacts, classes, sources, files or logs from upstream modules. Prior to running an application module, a set of "includes" would copy these things from various modules into the workspace.

An application module would be able to use a workspace (under a lock), from which it could read and write, and downstream modules would be able to access files created in the workspace like files in the repo.

The inputs to the workspace would be hashed before running the application module, and the contents of the workspace would be hashed on exit, and cached (along with STDOUT from the application module) if the module is marked as deterministic.

Further details TBC.

**Comment by coconutpalm** [comment]
I'd like to see classloader-level isolation for a given workspace/module version when running an application à la OSGI but without all the metadata ceremony OSGi has.

**Comment by propensive** [comment]
I experimented a bit with classloader isolation a while ago, so that we didn't need to launch a new JVM for every application module, but I didn't manage to get it working. I'm not convinced it's impossible, but it's not straightforward. Currently, we have (stronger!) JVM-level isolation, at the expense of having to launch a new JVM.

**Comment by coconutpalm** [comment]
I'd been meaning to write you about this because I've been dreaming of a
Scala runtime that works with Fury to keep the application running live by
hotswapping the latest code into a JVM as it's being developed.

It'd love to see the code you wrote because I've worked with this kind of
thing before.

Also, have you looked at OSGi implementations like equinox that already
implement this?


Dave

**Comment by propensive** [comment]
I haven't looked at Equinox for over a decade, but I did play around with it for a bit back then, though I seemed to spend most of my time generating manifest files...

But I'd love to talk some more. Do you have time for a call later this week or next? (And if so, what timezone are you in?)

**Comment by coconutpalm** [comment]
Yes let's talk!  I'm in cdt.  You can email privately at [david á coconut
hyphen palm hyphen software point côm].

Dave
